source/adapters/level_zero/queue.cpp

```diff
@@ -1697,14 +1697,7 @@ ur_result_t ur_queue_handle_t_::synchronize() {
     // zero handle can have device scope, so we can't synchronize the last
     // event.
     if (isInOrderQueue() && !LastCommandEvent->IsDiscarded) {
-      if (UrL0QueueSyncNonBlocking) {
-        auto SyncZeEvent = LastCommandEvent->ZeEvent;
-        this->Mutex.unlock();
-        ZE2UR_CALL(zeHostSynchronize, (SyncZeEvent));
-        this->Mutex.lock();
-      } else {
-        ZE2UR_CALL(zeHostSynchronize, (LastCommandEvent->ZeEvent));
-      }
+      ZE2UR_CALL(zeHostSynchronize, (LastCommandEvent->ZeEvent));
```

Comment:
Below this, there's `LastCommandEvent = nullptr;` assignment, which, if it was changed from another thread, will be overwritten. It may need to changed to something like:
```
auto savedLastEvent = LastCommandEvent
...
compare_and_swap(LastCommandEvent, savedLastEvent, null);
```
This way, this thread will only update the LastCommandEvent if nothing else has changed it.

Reply:
Should we be doing that in a different patch?  I can incorporate that into this patch, but that was not the original target. The purpose of this patch was to maintain the lock and call the host sync in this case.

Reply:
Up to you, I think this patch will fix in order queues, may still have a race in the out-of-order scenario. And it's possible that using CAS addresses both at the same time.

Reply:
Good idea, let me add the CAS and retest to ensure the original issue is still addressed.

Reply:
I just realized that we might not need a CAS here. This variable is always modified and read under a lock. 
So a simple `if (LastCommandEvent == savedLastEvent) LastCommandEvent = nullptr;` (with `savedLastEvent` assigned [here](https://github.com/oneapi-src/unified-runtime/pull/1749/files#diff-2c1af7ac7b1b3f44932c1e685b141784c2d14a66126ea7a383f9f30d919678f5L1695)) should do the trick.